components/inputs/input-date-range.js

```diff
@@ -129,16 +149,71 @@ class InputDateRange extends RtlMixin(LocalizeCoreElement(LitElement)) {
 		if (input) input.focus();
 	}
 
+	async validate(showErrors) {
+		const errors = await super.validate(showErrors);
+		if (errors.length !== 0) {
+			return errors;
+		}
+		setTimeout(async() => {
+			return Promise.all([
+				await this.shadowRoot.querySelector('.d2l-input-date-range-start').validate(showErrors),
+				await this.shadowRoot.querySelector('.d2l-input-date-range-end').validate(showErrors)]
+			).then((res) => {
+				return res.reduce((acc, errors) => [...acc, ...errors], []);
+			});
+		});
+	}
+
+	get validationMessageBadInput() {
+		return this.localize('components.input-date-range.errorBadInput', { startLabel: this._startLabel, endLabel: this._endLabel });
+	}
+
+	get _endLabel() {
+		return this.endLabel ? this.endLabel : this.localize('components.input-date-range.endDate');
+	}
+
 	async _handleChange(e) {
-		// TODO: validation that start is before end
 		const elem = e.target;
-		if (elem.classList.contains('d2l-input-date-range-start')) this.startValue = elem.value;
-		else this.endValue = elem.value;
+		if (elem.classList.contains('d2l-input-date-range-start')) {
+			this.startValue = elem.value;
+		} else {
+			this.endValue = elem.value;
+		}
+		await this.requestValidate();
 		this.dispatchEvent(new CustomEvent(
 			'change',
 			{ bubbles: true, composed: false }
 		));
 	}
 
+	_handleDropdownOpen(e) {
+		const id = e.target.id;
+		const tooltip = this.shadowRoot.querySelector(`d2l-tooltip[for="${id}"]`);
+		if (tooltip && tooltip.showing) tooltip.hide();
+	}
+
+	_handleInnerFocus(e) {
+		// in order for calendar close to cause tooltip to show
+		const id = e.target.id;
+		const tooltip = this.shadowRoot.querySelector(`d2l-tooltip[for="${id}"]`);
+		if (tooltip && !tooltip.showing) tooltip.show();
+	}
```

Comment:
I assume in this case the tooltip's target already has focus but the tooltip doesn't open because it got focus before the tooltip was added? If thats the case this feels like something that should be fixed in `d2l-tooltip` so it starts opened when added to a target that already has focus or hover.

Reply:
So the tooltip's target is the `input-date` but when the calendar closes, the focus moves to the `d2l-input-text` within the `input-date` which does not trigger focus up to the `input-date`

Reply:
I do think we can likely make this work with tooltip automatically since the `activeElement` will be reported as the `input-date` even if its the `input-text` with focus. Is this code specifically to hand the case where they click a date triggering an invalid range and the calendar closes or are there other cases it handles?

Reply:
Okay yeah if we can handle that without this then that would be good.

Yup, basically, it's for the case where the end date is before start date and then the calendar is open and then closes (could be that the selected dates were bad before opening and they did not fix it while the calendar was open, and so it's still a bad input when the calendar closes).

Reply:
This _should_ now automatically be handled by tooltip, let me know if you run into any issues.